### PR TITLE
search: Restructure "sync query from URL" logic

### PR DESCRIPTION
**Pull request by fkling**
**tl;dr**: Instead of relying on a couple of `useEffect`s in multiple components
to sync/update the search query from the URL we are now listening to URL
changes in the app root and update the store directly from there.

---

It's possible that in some obscure situations the React effect for updating
the query state doesn't happen. But there is also a more general issue
with how we sync the query information from the URL to the app state,
which causes multiple re-renders and "flashing" of the query in the
search query input.

The "flashing" happens because we use `useEffect` in a lot of places to 
trigger updates to the search query store. `useEffect` callbacks are
executed *after* the component has already rendered, which is too late.

I tried various things moving the "URL to store" syncing logic to various
places, but all of them had some issues. The two main issues are:
- Query information is used in multiple places, so ideally the store is
  update before any component that depends on it is rendered. That means
  a store update should be triggered close to the root component.
- When syncing is tried to component re-rendering we restrict ourselves to
  updating the store *after* the containing component renders. We cannot
  update the store *while* the component re-renders because that would cause
  other components to update which causes a React error.

Eventually I ended up attaching a handler that listens to location changes
in the web app root and updates the query store accordingly, which I think
makes the most sense, at least for the time being. This causes the store to
be updated independently of component re-rendering.
The logic for updating the query and the search context, which was previously
contained in `Layout.tsx` and `GlobalNavbar.tsx`, is now located in a single
place.

Before/after video: 

https://user-images.githubusercontent.com/179026/189132525-c4409f68-48c8-4c04-b4d9-f0825fdd8c9d.mp4

## Test plan

To (not) reproduce the "empty input bug":
- Open Sourcegraph in a window that's < 1000px wide. It should show the collapsed
  menu bar.
- Increase size to see full menu bar.
- Enter and submit search query on search homepage
- The query input should contain the query on the results page

`index.test.ts` (which was renamed from `index.test.tsx`) contains additional
unit tests.

## App preview:

- [Web](https://sg-web-fkling-query-state-from-url.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-udazqonxlj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/search/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+})
+
+describe('repoFilterForRepoRevision escapes values with spaces', () => {
+    test('escapes spaces in value', () => {
+        expect(repoFilterForRepoRevision('7 is my final answer', false)).toMatchInlineSnapshot(
+            '"^7\\\\ is\\\\ my\\\\ final\\\\ answer$"'
+        )
+    })
+})
+
```
Comment (fkling) [inline client/web/src/search/index.test.ts:156]: This is all existing code from the original `index.test.tsx` file (I renamed the file to `index.test.ts` because it didn't contain any JSX).

File: client/web/src/search/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                .pipe(
+                    last(),
+                    tap(() => {
+                        expect(useNavbarQueryState.getState().queryState.query).toBe('context:me test')
+                    })
+                )
+                .toPromise()
+        })
+    })
+})
```
Comment (fkling) [inline client/web/src/search/index.test.ts:321]: This are new tests / tests migrated from `Layout.test.tsx`.

File: client/web/src/search/index.tsx [hunk 3]
```diff
@@ -189,3 +195,86 @@ export interface SearchStreamingProps {
...
+
+            const queryClean =
+                // If a global search context spec is available to the user, we omit it from the
+                // query and move it to the search contexts dropdown
+                globalSearchContextSpec && isSearchContextAvailable && showSearchContext
+                    ? omitFilter(query, globalSearchContextSpec.filter)
+                    : query
+
+            // Update global query state
+            useNavbarQueryState.setState(({ searchPatternType, searchCaseSensitivity }) => ({
```
Comment (novoselrok) [inline client/web/src/search/index.tsx:268]: Nit: it feels a bit weird to have a `setState` side effect inside of a `map` function. What do you think about returning the state alongside the `parsedSearchURL` and `searchContextSpec` and then calling `setState` inside the `.subscribe` callback in `SourcegraphWebApp`?

Comment (fkling) [reply client/web/src/search/index.tsx:268]: Yeah, that would be my preferred solution as well but I wasn't sure how to make this work with the tests... but now that I have a working solution, let me look into adding back some form of the `setQueryStateFromURL` function.

FWIW the `.map` is only used as a simple solution to return observable that's connected to the input observables. I could have also subscribed inside the function and return a new observable and hook up the subscriptions somehow. This was easier (although more confusing for sure).

Comment (fkling) [reply client/web/src/search/index.tsx]: Updated.

**Review by novoselrok — APPROVED** [review]
LGTM, fixes the issue for me locally.

File: client/web/src/stores/navbarSearchQueryState.ts [hunk 4]
```diff
@@ -59,9 +59,13 @@ export function setSearchCaseSensitivity(searchCaseSensitivity: boolean): void {
...
- * Update or initialize query state related data from URL search parameters
+ * Update or initialize query state related data from URL search parameters.
+ *
+ * @param parsedSearchURL contains the information extracted from a URL
+ * @param query can be used to specify the query to use when it differs from
+ * the one contained in the URL (e.g. when the context:... filter got removed)
  */
-export function setQueryStateFromURL(urlParameters: string): void {
+export function setQueryStateFromURL(parsedSearchURL: ParsedSearchURL, query = parsedSearchURL.query ?? ''): void {
     if (useNavbarQueryState.getState().parametersSource > InitialParametersSource.URL) {
```
Comment (vovakulikov) [inline client/web/src/stores/navbarSearchQueryState.ts:69]: @fkling, why do we need this line? it seems that none of `parametersSource` modes could be greater than the URL,  in this case, this check never passes

File: client/web/src/SourcegraphWebApp.tsx [hunk 5]
```diff
@@ -299,6 +299,37 @@ export class SourcegraphWebApp extends React.Component<
             console.error('Error sending search context to extensions!', error)
         })
 
+        // Update search query state whenever the URL changes
+        this.subscriptions.add(
+            getQueryStateFromLocation({
```
Comment (vovakulikov) [inline client/web/src/SourcegraphWebApp.tsx:304]: @fkling, I'm curious about the general approach here. 

We use the URL as the main place where the query lives. And then, we listen to changes from the URL in order to update the global store value. It feels (maybe this false concern) that the dependence tree here should be reversed. When we update the search and hit submit, we should write to the global store a new submitted value, and then we will update the ULR query params and URL-related effects with Zustand URL middleware. 

This may introduce a few problems if the course, for example, What will we do with native or react-router links that change queries, and how can they work in this system? 

Currently, it's very difficult to keep this query life circle in mind while working with a query global state.

File: client/web/src/stores/navbarSearchQueryState.ts (hunk 4 above)
Comment (fkling) [reply client/web/src/stores/navbarSearchQueryState.ts:69]: I merely added that for consistency should we ever introduce a source with higher priority than URL.

File: client/web/src/SourcegraphWebApp.tsx (hunk 5 above)
Comment (fkling) [reply client/web/src/SourcegraphWebApp.tsx:304]: Yeah, I'm torn but still leaning towards having the data flow downstream from the URL, also because, as you mentioned, it makes working with links easier.

I think the data flow is actually OK if you accept the URL always being the source of truth:

````
URL --- changes --> Store -----> Input
 ^                    ^-- edit ----| |
 |------------ submit ---------------|
```

The store is seeded from the URL, directly updated (via input or other callsites) and submitting a query (which can happen in different ways too) updates the URL.

I think the main problem so far was that the logic was too coupled with the React render life cycle and spread over more or less arbitrary components. I hope the new setup makes it easier to follow the data.

Having said that, we might reach a point in the future where we can't or don't want to use the URL as source of truth for the search/application state. Then it it could make sense to revisit this.
